File: Difficulty: Medium/Rat in a Maze/rat-in-a-maze.cpp
class Solution {
  public:
    vector<string> ratInMaze(vector<vector<int>>& maze) {
        // code here
        vector<string> ans;
        vector<vector<int>> dp(maze.size(),vector<int>(maze.size(),0));
        helper(maze,ans,dp,"",0,0);
        return ans;
    }
    void helper(vector<vector<int>>& maze, vector<string>& ans, vector<vector<int>>& dp, string s,int i,int j){
        if(i==maze.size()-1 && j==maze.size()-1){
            ans.push_back(s);
            return;
        }
        dp[i][j] = 1;
        if(isSafe(maze,dp,i+1,j))
            helper(maze,ans,dp,s+'D',i+1,j);
        if(isSafe(maze,dp,i,j-1))
            helper(maze,ans,dp,s+'L',i,j-1);
        if(isSafe(maze,dp,i,j+1))
            helper(maze,ans,dp,s+'R',i,j+1);
        if(isSafe(maze,dp,i-1,j))
            helper(maze,ans,dp,s+'U',i-1,j);
        dp[i][j]=0;
    }
    bool isSafe(vector<vector<int>>& maze, vector<vector<int>>& dp,int i,int j){
        if(i<maze.size() && j<maze.size() && i>=0 && j>=0 && maze[i][j]==1 && dp[i][j]==0)
            return true;
        return false;
    }
};

File: Difficulty: Medium/Rat in a Maze/rat-in-a-maze.java
class Solution {
    private boolean isSafe(int[][] maze, int[][] visited,int i, int j){
        if(i<maze.length && j<maze.length && i>=0 && j>=0 && maze[i][j]==1 && visited[i][j]==-1)
            return true;
        return false;
    }
    private void helper(int[][] maze, int[][] visited, ArrayList<String> ans, String s, int i, int j){
        if(i==maze.length-1 && j==maze.length-1){
            ans.add(s);
            return;
        }
        visited[i][j]=1;
        if(isSafe(maze,visited,i+1,j))
            helper(maze,visited,ans,s+"D",i+1,j);
        if(isSafe(maze,visited,i,j-1))
            helper(maze,visited,ans,s+"L",i,j-1);
        if(isSafe(maze,visited,i,j+1))
            helper(maze,visited,ans,s+"R",i,j+1);
        if(isSafe(maze,visited,i-1,j))
            helper(maze,visited,ans,s+"U",i-1,j);
        visited[i][j]=-1;
    }
    public ArrayList<String> ratInMaze(int[][] maze) {
        // code here
        int[][] visited = new int[maze.length][maze.length];
        for(int i=0;i<maze.length;i++){
            Arrays.fill(visited[i],-1);
        }
        ArrayList<String> ans = new ArrayList<>();
        helper(maze,visited,ans,"",0,0);
        return ans;
    }
}